library/macos/FLETextInputPlugin.mm

```diff
@@ -0,0 +1,390 @@
+ * The FLEViewController to manage input for.
+ */
+@property(nonatomic, weak) FLEViewController *flutterViewController;
+
+/**
+ * Handles a Flutter system message on the text input channel.
+ */
+- (void)handleMethodCall:(FLEMethodCall *)call result:(FLEMethodResult)result;
+
+@end
+
+@implementation FLETextInputPlugin
+
+- (instancetype)initWithViewController:(FLEViewController *)viewController {
+  self = [super init];
+  if (self != nil) {
+    _flutterViewController = viewController;
+    _channel = [FLEMethodChannel methodChannelWithName:kTextInputChannel
+                                       binaryMessenger:viewController
+                                                 codec:[FLEJSONMethodCodec sharedInstance]];
+    __weak FLETextInputPlugin *weakSelf = self;
+    [_channel setMethodCallHandler:^(FLEMethodCall *call, FLEMethodResult result) {
+      [weakSelf handleMethodCall:call result:result];
+    }];
+    _textInputContext = [[NSTextInputContext alloc] initWithClient:self];
+  }
+  return self;
+}
+
+#pragma mark - Private
+/**
+ * Converts an NSDictionary to a State object. Will raise an exception if the dictionary is not
+ * configured properly.
+ */
+- (flutter_desktop_embedding::State)stateFromDictionary:(NSDictionary *)dict {
+  flutter_desktop_embedding::State new_state;
+  NSString *text = dict[kTextKey];
+  if (!text) {
+    throw std::invalid_argument("Set editing state has been invoked, but without text.");
+  }
+  new_state.text = [text UTF8String];
+
+  NSNumber *selection_base = dict[kSelectionBaseKey];
+  NSNumber *selection_extent = dict[kSelectionExtentKey];
+  if (!selection_base || !selection_extent) {
+    throw std::invalid_argument("Selection base/extent values invalid.");
+  }
+  new_state.selection.begin = [selection_base unsignedLongValue];
+  new_state.selection.end = [selection_extent unsignedLongValue];
+
+  NSNumber *composing_base = dict[kComposingBaseKey];
+  NSNumber *composing_extent = dict[kComposingExtentKey];
+
+  new_state.composing.begin =
+      composing_base ? [composing_base unsignedLongValue] : new_state.selection.begin;
+  new_state.composing.end =
+      composing_extent ? [composing_extent unsignedLongValue] : new_state.selection.end;
+
+  NSString *text_affinity = dict[kSelectionAffinityKey];
+  new_state.text_affinity = text_affinity ? [text_affinity UTF8String] : kTextAffinityDownstream;
+
+  new_state.isDirectional = [dict[kSelectionIsDirectionalKey] boolValue];
+
+  return new_state;
+}
+
+/**
+ * Creates a dictionary from a given State;
+ */
+- (NSDictionary *)dictionaryFromState:(flutter_desktop_embedding::State)state {
+  std::string affinity = state.text_affinity.compare(kTextAffinityUpstream) == 0
+                             ? kTextAffinityUpstream
+                             : kTextAffinityDownstream;
+  NSString *affinityString = [NSString stringWithCString:affinity.c_str()
+                                                encoding:[NSString defaultCStringEncoding]];
```

Comment:
`defaultCStringEncoding` is almost never what you want, since you don't know what you'll get. Use `UTF8String` (or if you know Flutter is using another encoding in this channel—which seems unlikely—use that explicitly).